File: Javascript/Two Pointers/M-424. Longest Repeating Character Replacement.ts
function characterReplacement(s: string, k: number): number {
  const charFreq = {};
  let longest = 0;
  let curLongest = 0;
  let l = 0;
  let r = 0;

  while (r < s.length) {
    charFreq[s[r]] = (charFreq[s[r]] || 0) + 1;
    curLongest = Math.max(curLongest, charFreq[s[r]]);

    while (r - l + 1 - curLongest > k) {
      charFreq[s[l]] -= 1;
      l++;
    }

    longest = Math.max(longest, r - l + 1);
    r++;
  }

  return longest;
}

const testData = [
  {
    s: 'ABAB',
    k: 2,
    ans: 4,
  },
  {
    s: 'AABABBA',
    k: 1,
    ans: 4,
  },
];

for (const data of testData) {
  console.log(data.ans === characterReplacement(data.s, data.k));
}

File: Two Pointers/M-424. Longest Repeating Character Replacement/M-424. Longest Repeating Character Replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        char_freq = {}
        longest, cur_longest = 0, 0
        l, r = 0, 0

        while r < len(s):
            char_freq[s[r]] = (char_freq.get(s[r]) or 0) + 1
            cur_longest = max(cur_longest, char_freq[s[r]])

            while r - l + 1 - cur_longest > k:
                char_freq[s[l]] -= 1
                l += 1

            longest = max(longest, r - l + 1)
            r += 1

        return longest
